Client/Frontend/AuthenticationManager/PasscodeViewController.swift

```diff
@@ -0,0 +1,297 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import SnapKit
+
+private struct PasscodeUX {
+    static let TitleVerticalSpacing: CGFloat = 32
+    static let DigitSize: CGFloat = 30
+    static let TopMargin: CGFloat = 80
+    static let PaneSwipeDuration: NSTimeInterval = 0.3
+    static let PasscodeFieldSize: CGSize = CGSize(width: 160, height: 32)
+}
+
+@objc protocol PasscodeInputViewDelegate: class {
+    func passcodeInputView(inputView: PasscodeInputView, didFinishEnteringCode code: String)
+}
+
+/// A custom, keyboard-able view that displays the blank/filled digits when entrering a passcode.
+class PasscodeInputView: UIView, UIKeyInput {
+    weak var delegate: PasscodeInputViewDelegate?
+
+    var digitFont: UIFont = UIConstants.PasscodeEntryFont
+
+    var blankCharacter: Character = "-"
+
+    var filledCharacter: Character = "•"
+
+    private let passcodeSize: Int
+
+    private var inputtedCode: String = ""
+
+    private var blankDigitString: NSAttributedString {
+        return NSAttributedString(string: "\(blankCharacter)", attributes: [NSFontAttributeName: digitFont])
+    }
+
+    private var filledDigitString: NSAttributedString {
+        return NSAttributedString(string: "\(filledCharacter)", attributes: [NSFontAttributeName: digitFont])
+    }
+
+    @objc var keyboardType: UIKeyboardType = .NumberPad
+
+    init(frame: CGRect, passcodeSize: Int) {
+        self.passcodeSize = passcodeSize
+        super.init(frame: frame)
+        opaque = false
+    }
+
+    convenience init(passcodeSize: Int) {
+        self.init(frame: CGRectZero, passcodeSize: passcodeSize)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func canBecomeFirstResponder() -> Bool {
+        return true
+    }
+
+    func resetCode() {
+        inputtedCode = ""
+        setNeedsDisplay()
+    }
+
+    @objc func hasText() -> Bool {
+        return inputtedCode.characters.count > 0
+    }
+
+    @objc func insertText(text: String) {
+        guard inputtedCode.characters.count < passcodeSize else {
+            return
+        }
+
+        inputtedCode += text
+        setNeedsDisplay()
+        if inputtedCode.characters.count == passcodeSize {
+            delegate?.passcodeInputView(self, didFinishEnteringCode: inputtedCode)
+            resignFirstResponder()
+        }
+    }
+
+    @objc func deleteBackward() {}
```

Comment:
Why do you need an empty override here? Maybe add a comment above explaining what this is for.